### PR TITLE
Replace and Deprecate TS_VERIFY_CTX Functions and Add Documentation

### LINKED ISSUE #18854: Leaks or use-after-frees in TS_VERIFY_CTX accessors
The `TS_VERIFY_CTX` accessors return the value passed in. This is a bit strange.

https://github.com/openssl/openssl/blob/cbb1cda67f61b34c89fb49c8e2267ec6217bc33f/crypto/ts/ts_verify_ctx.c#L51-L77

One possible explanation is that the appropriate freeing of already present struct members were missed. This is the interpretation taken in #16347 for `TS_VERIFY_CTX_set_imprint()`.

Or perhaps the idea of https://github.com/openssl/openssl/commit/ca4a494cb7437a7af155361b0084de8329c0bf25 was that they should return the old value of the relevant `ctx` member so that the caller can free them? However, that doesn't make sense with the error checking that was added with their use in ts.c.

Either way, the current behavior means that whatever was set on the `ctx` is likely leaked if an application calls one of these setters a second time.

As already mentioned, this was changed for `TS_VERIFY_CTX_set_imprint()` in #16347 (without treating the other functions the same way). This was a dangerous change since it makes the API behavior inconsistent and, worse, it causes a use after free or a double free for callers who would previously take care of freeing the old pointer themselves to avoid a leak.

This problem would arise for example in [ruby/openssl][1] if someone were to add an `X509_STORE_free(store)` at the top of `TS_VERIFY_CTX_set_store()` similar to #16347 .

Presumably the only way to fix this without breaking existing consumers is to deprecate the existing API and to add `set0` or `set1` versions of these setters that don't have these issues.

[1]: https://github.com/ruby/openssl/blob/b31446464e1e9f8bd0c58e92f9e74a9c7663b0e0/ext/openssl/ossl_ts.c#L831-L839

### PR DESCRIPTION
**Pull request by erbsland-dev**
This pull request addresses two main areas related to the `TS_VERIFY_CTX` functions:

## 1. Replace and Deprecate Functions

Fixes #18854 by replacing and deprecating the following functions due to poorly documented memory handling, which could lead to memory leaks:
- `TS_VERIFY_CTX_set_data`
- `TS_VERIFY_CTX_set_store`
- `TS_VERIFY_CTX_set_certs`
- `TS_VERIFY_CTX_set_imprint`

These functions are replaced with:
- `TS_VERIFY_CTX_set0_data`
- `TS_VERIFY_CTX_set0_store`
- `TS_VERIFY_CTX_set0_certs`
- `TS_VERIFY_CTX_set0_imprint`

Updates have been made to existing code to use the new function calls instead of the deprecated ones.

### 2. Add and Update Documentation
- Mark the existing `TS_VERIFY_CTX_set_certs` function as deprecated in the documentation.
- Add missing documentation for the deprecated functions:
  - `TS_VERIFY_CTX_set_data`
  - `TS_VERIFY_CTX_set_imprint`
  - `TS_VERIFY_CTX_set_store`
- Write missing documentation for the following functions:
  - `TS_VERIFY_CTX_new`
  - `TS_VERIFY_CTX_init`
  - `TS_VERIFY_CTX_free`
  - `TS_VERIFY_CTX_cleanup`
  - `TS_VERIFY_CTX_set_flags`
  - `TS_VERIFY_CTX_add_flags`
  - `TS_VERIFY_CTX_set0_data`
  - `TS_VERIFY_CTX_set0_imprint`
  - `TS_VERIFY_CTX_set0_store`
  - `TS_VERIFY_CTX_set0_certs`

These changes improve memory management, enhance future extensibility, and provide clearer usage guidelines through updated and additional documentation.

- [x] documentation is added or updated

### CODE REVIEW

**Comment by nhorman** [comment]
@erbsland-dev looks like there are a few minor doc nits to correct

**Comment by erbsland-dev** [comment]
@nhorman I fixed the documentation issues.

**Review by nhorman — APPROVED** [review]

**Review by shahsb — APPROVED** [review]

File: doc/man3/TS_VERIFY_CTX.pod [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+TS_VERIFY_CTX_set_store() is used to set the certificate store. A previously
+assigned store is B<not freed> by this call. B<s> is the store to assign.
+
+TS_VERIFY_CTX_set_certs() is used to set the server's certificate chain.
+A previously assigned stack is B<not freed> by this call. B<certs> is a stack
+of B<X509> certificates.
+
+TS_VERIFY_CTS_set_certs() is a misspelled version of TS_VERIFY_CTX_set_certs()
+which takes the same parameters and returns the same result.
+
```
Comment (t8m) [inline doc/man3/TS_VERIFY_CTX.pod:112]: Nit: please drop this extra empty line

File: doc/man3/TS_VERIFY_CTX.pod [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+                                          unsigned char *hexstr, long len);
+ X509_STORE *TS_VERIFY_CTX_set_store(TS_VERIFY_CTX *ctx, X509_STORE *s);
+ STACK_OF(X509) *TS_VERIFY_CTX_set_certs(TS_VERIFY_CTX *ctx,
+                                         STACK_OF(X509) *certs);
+
+The following function has been deprecated since OpenSSL 3.0:
+
+ STACK_OF(X509) *TS_VERIFY_CTS_set_certs(TS_VERIFY_CTX *ctx,
+                                         STACK_OF(X509) *certs);
+
```
Comment (t8m) [inline doc/man3/TS_VERIFY_CTX.pod:43]: Please drop this empty line

File: include/openssl/ts.h [hunk 3]
```diff
@@ -418,14 +418,31 @@ void TS_VERIFY_CTX_free(TS_VERIFY_CTX *ctx);
 void TS_VERIFY_CTX_cleanup(TS_VERIFY_CTX *ctx);
 int TS_VERIFY_CTX_set_flags(TS_VERIFY_CTX *ctx, int f);
 int TS_VERIFY_CTX_add_flags(TS_VERIFY_CTX *ctx, int f);
+# ifndef OPENSSL_NO_DEPRECATED_3_4
+OSSL_DEPRECATEDIN_3_4_FOR("Previous ctx data was not released, replace with TS_VERIFY_CTX_set0_data().")
 BIO *TS_VERIFY_CTX_set_data(TS_VERIFY_CTX *ctx, BIO *b);
+# endif
+int TS_VERIFY_CTX_set0_data(TS_VERIFY_CTX *ctx, BIO *b);
+# ifndef OPENSSL_NO_DEPRECATED_3_4
+OSSL_DEPRECATEDIN_3_4_FOR("Memory usage was not defined, replace with TS_VERIFY_CTX_set0_imprint().")
```
Comment (t8m) [inline include/openssl/ts.h:427]: I'd suggest not putting the deprecation reason into the message. Or at least make it shorter `Unclear semantics, replace with...`.

File: include/openssl/ts.h [hunk 4]
```diff
@@ -418,14 +418,31 @@ void TS_VERIFY_CTX_free(TS_VERIFY_CTX *ctx);
...
 unsigned char *TS_VERIFY_CTX_set_imprint(TS_VERIFY_CTX *ctx,
                                          unsigned char *hexstr, long len);
+# endif
+int TS_VERIFY_CTX_set0_imprint(TS_VERIFY_CTX *ctx,
+                               unsigned char *hexstr, long len);
+# ifndef OPENSSL_NO_DEPRECATED_3_4
+OSSL_DEPRECATEDIN_3_4_FOR("Previous ctx data was not released, replace with TS_VERIFY_CTX_set0_store().")
```
Comment (t8m) [inline include/openssl/ts.h:434]: As above

File: include/openssl/ts.h [hunk 5]
```diff
@@ -418,14 +418,31 @@ void TS_VERIFY_CTX_free(TS_VERIFY_CTX *ctx);
...
 X509_STORE *TS_VERIFY_CTX_set_store(TS_VERIFY_CTX *ctx, X509_STORE *s);
+# endif
+int TS_VERIFY_CTX_set0_store(TS_VERIFY_CTX *ctx, X509_STORE *s);
 # ifndef OPENSSL_NO_DEPRECATED_3_0
 #  define TS_VERIFY_CTS_set_certs(ctx, cert) TS_VERIFY_CTX_set_certs(ctx,cert)
 # endif
+# ifndef OPENSSL_NO_DEPRECATED_3_4
+OSSL_DEPRECATEDIN_3_4_FOR("Previous ctx data was not released, replace with TS_VERIFY_CTX_set0_certs().")
```
Comment (t8m) [inline include/openssl/ts.h:442]: As above

File: doc/man3/TS_VERIFY_CTX.pod [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+time. TSP defines a Time Stamping Authority (TSA) and an entity that makes
+requests to the TSA. Usually, the TSA is referred to as the server side, and the
+requesting entity is referred to as the client.
+
+In TSP, when a server sends a response to a client, the server normally
+needs to sign the response data - the TimeStampToken (TST) - with its private
+key. Then the client verifies the received TST using the server's certificate
+chain.
+
+For all the following methods, unless noted otherwise, B<ctx> is the
```
Comment (t8m) [inline doc/man3/TS_VERIFY_CTX.pod:58]: I know you're just copying the page however as you're already touching this, could you please use `I<ctx>` and similarly elsewhere throughout the manpage where names of arguments are referenced?

**Review by t8m — CHANGES_REQUESTED** [review]
Good work!

**Comment by t8m** [comment]
(Tests exempted because we do not have any TS subsystem testcases at all yet.)

**Comment by t8m** [comment]
@erbsland-dev Could you please also rebase this to resolve the conflict?

**Comment by erbsland-dev** [comment]
@t8m Sure, I'll do it, but it probably won't be until the weekend.

**Comment by erbsland-dev** [comment]
@t8m I just rebased the pull request to the latest master. Could you please review how I merged `util/libcrypto.num`? I’m not entirely sure I did it correctly.

**Comment by erbsland-dev** [comment]
@t8m Rebasing done and I integrated all your change requests.

**Review by t8m — APPROVED** [review]
@nhorman please reconfirm your approval

**Comment by nhorman** [comment]
approval holds

**Comment by t8m** [comment]
Merged to the master branch after reformatting the commit messages. Thank you for your contribution.

**Pull request closed without merging**